subprojects/dependency-management/src/test/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformProjectBuilderTest.groovy

```diff
@@ -0,0 +1,177 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.integtests.resolve.transform
+
+import com.google.common.collect.ImmutableList
+import org.gradle.api.artifacts.transform.InputArtifact
+import org.gradle.api.artifacts.transform.TransformAction
+import org.gradle.api.artifacts.transform.TransformOutputs
+import org.gradle.api.artifacts.transform.TransformParameters
+import org.gradle.api.attributes.Attribute
+import org.gradle.api.file.FileSystemLocation
+import org.gradle.api.internal.artifacts.VariantTransformRegistry
+import org.gradle.api.internal.artifacts.ivyservice.resolveengine.artifact.ResolvableArtifact
+import org.gradle.api.internal.artifacts.transform.DefaultTransformUpstreamDependenciesResolver
+import org.gradle.api.internal.artifacts.transform.TransformationStep
+import org.gradle.api.internal.artifacts.transform.TransformationSubject
+import org.gradle.api.internal.artifacts.transform.TransformerInvocationFactory
+import org.gradle.api.provider.Provider
+import org.gradle.api.tasks.Classpath
+import org.gradle.internal.Try
+import org.gradle.internal.component.local.model.OpaqueComponentArtifactIdentifier
+import org.gradle.internal.execution.fingerprint.InputFingerprinter
+import org.gradle.test.fixtures.AbstractProjectBuilderSpec
+import org.gradle.test.fixtures.file.TestFile
+
+import java.util.concurrent.atomic.AtomicInteger
+
+class ArtifactTransformProjectBuilderTest extends AbstractProjectBuilderSpec {
+    public static final AtomicInteger INVOCATION_COUNT = new AtomicInteger(0)
+    public static final String SELECTED_PATH = "selected.txt"
+
+    def artifactType = Attribute.of('artifactType', String)
+
+    def setup() {
+        INVOCATION_COUNT.set(0)
+    }
+
+    def "input artifact selection is restored when using the in-memory cache"() {
+        def transformer = registerTransform(IdentityTransform)
+
+        def inputArtifact1 = file("input1.txt")
+        inputArtifact1.text = "Hello"
+
+        def inputArtifact2 = file("input2.txt")
+        inputArtifact2.text = "Hello"
+
+        expect:
+        invoke(transformer, inputArtifact1).get() == [inputArtifact1]
+        invoke(transformer, inputArtifact2).get() == [inputArtifact2]
+        INVOCATION_COUNT.get() == 1
+    }
+
+    def "input artifact paths are restored when using the in-memory cache"() {
+        def transformer = registerTransform(SelectFileTransform)
+
+        def inputArtifact1 = file("input1")
+
+        def selectedFile1 = inputArtifact1.file(SELECTED_PATH)
+        selectedFile1.text = "Hello"
+
+        def inputArtifact2 = file("input2.txt")
+
+        def selectedFile2 = inputArtifact2.file(SELECTED_PATH)
+        selectedFile2.text = "Hello"
+
+        expect:
+        invoke(transformer, inputArtifact1).get() == [selectedFile1]
```

Comment:
Reading this test I think we could make it a bit more clear that `invoke()` is something relevant to this test only, not something that comes from a superclass for example. Say, by calling it `transform()` or `invokeTransform()`.

Reply:
I renamed the method to `invokeTransform`.